qa/tasks/keycloak.py

```diff
@@ -0,0 +1,331 @@
+    """
+    Downloading the two required tar files
+    1. Keycloak
+    2. Wildfly (Application Server)
+    """
+    assert isinstance(config, dict)
+    log.info('Installing packages for Keycloak...')
+
+    for (client, _) in config.items():
+        (remote,) = ctx.cluster.only(client).remotes.keys()
+        test_dir=teuthology.get_testdir(ctx)
+        toxvenv_sh(ctx, remote, ['wget', 'https://downloads.jboss.org/keycloak/11.0.0/keycloak-11.0.0.tar.gz'])
+        toxvenv_sh(ctx, remote, ['tar', '-C', test_dir, '-xvzf', 'keycloak-11.0.0.tar.gz'])
+
+        tmp_dir=os.path.join(test_dir,"keycloak-11.0.0/")
+        
+        toxvenv_sh(ctx, remote, ['cd', tmp_dir, run.Raw('&&'), 'wget', 'https://downloads.jboss.org/keycloak/11.0.0/adapters/keycloak-oidc/keycloak-wildfly-adapter-dist-11.0.0.tar.gz'])
+        toxvenv_sh(ctx, remote, ['tar', '-C', tmp_dir, '-xvzf', '{tdr}/keycloak-wildfly-adapter-dist-11.0.0.tar.gz'.format(tdr=tmp_dir)])
+
+    try:
+        yield
+    finally:
+        log.info('Removing packaged dependencies of Keycloak...')
+        test_dir=teuthology.get_testdir(ctx)
+        tmp_dir=os.path.join(test_dir,"keycloak-11.0.0/")
+        for client in config:
+            ctx.cluster.only(client).run(
+                args=['rm', '-rf', tmp_dir],
+            )
+
+@contextlib.contextmanager
+def build(ctx,config):
+    """
+    Build process which needs to be done before starting a server.
+    """
+    assert isinstance(config, dict)
+    log.info('Building Keycloak...')
+    for (client,_) in config.items():
+        run_in_keycloak_dir(ctx, client, ['cd', 'bin', run.Raw('&&'), './jboss-cli.sh', '--file=adapter-elytron-install-offline.cli'])
+    try:
+        yield
+    finally:
+        pass
+
+@contextlib.contextmanager
+def run_keycloak(ctx,config):
+    """
+    This includes two parts:
+    1. Adding a user to keycloak which is actually used to log in when we start the server and check in browser.
+    2. Starting the server.
+    """
+    assert isinstance(config, dict)
+    log.info('Bringing up Keycloak...')
+    for (client,_) in config.items():
+        (remote,) = ctx.cluster.only(client).remotes.keys()
+        
+        ctx.cluster.only(client).run(
+            args=[
+                '{tdir}/bin/add-user-keycloak.sh'.format(tdir=get_keycloak_dir(ctx)),
+                '-r', 'master',
+                '-u', 'admin',
+                '-p', 'admin',
+            ],
+        )
+
+        toxvenv_sh(ctx, remote, ['cd', '{tdir}/bin'.format(tdir=get_keycloak_dir(ctx)), run.Raw('&&'), './standalone.sh', run.Raw('&'), 'exit'])
+    try:
+        yield
+    finally:
+        log.info('Stopping Keycloak Server...')
+
+        for (client, _) in config.items():
+            (remote,) = ctx.cluster.only(client).remotes.keys()
+            toxvenv_sh(ctx, remote, ['cd', '{tdir}/keycloak-11.0.0/bin'.format(tdir=teuthology.get_testdir(ctx)), run.Raw('&&'), './jboss-cli.sh', '--connect', 'command=:shutdown'])
```

Comment:
Here is another example of the keycloak version being hard coded that needs to be changed.